src/storage/partial_block_manager.cpp

```diff
@@ -0,0 +1,81 @@
+#include "duckdb/storage/partial_block_manager.hpp"
+
+namespace duckdb {
+
+//===--------------------------------------------------------------------===//
+// Partial Blocks
+//===--------------------------------------------------------------------===//
+PartialBlockAllocation PartialBlockManager::GetBlockAllocation(uint32_t segment_size) {
+	PartialBlockAllocation allocation;
+	allocation.block_manager = &block_manager;
+	allocation.allocation_size = segment_size;
+
+	// if the block is less than 80% full, we consider it a "partial block"
+	// which means we will try to fit it with other blocks
+	// check if there is a partial block available we can write to
+	if (segment_size <= free_space_threshold && GetPartialBlock(segment_size, allocation.partial_block)) {
+		//! there is! increase the reference count of this block
+		allocation.partial_block->state.block_use_count += 1;
+		allocation.state = allocation.partial_block->state;
+		block_manager.IncreaseBlockReferenceCount(allocation.state.block_id);
+	} else {
+		// full block: get a free block to write to
+		AllocateBlock(allocation.state, segment_size);
+	}
+	return allocation;
+}
+
+void PartialBlockManager::AllocateBlock(PartialBlockState &state, uint32_t segment_size) {
+	D_ASSERT(segment_size <= Storage::BLOCK_SIZE);
+	state.block_id = block_manager.GetFreeBlockId();
+	state.block_size = Storage::BLOCK_SIZE;
+	state.offset_in_block = 0;
+	state.block_use_count = 1;
+}
+
+bool PartialBlockManager::GetPartialBlock(idx_t segment_size, unique_ptr<PartialBlock> &partial_block) {
+	auto entry = partially_filled_blocks.lower_bound(segment_size);
+	if (entry == partially_filled_blocks.end()) {
+		return false;
+	}
+	// found a partially filled block! fill in the info
+	partial_block = move(entry->second);
+	partially_filled_blocks.erase(entry);
+
+	D_ASSERT(partial_block->state.offset_in_block > 0);
+	D_ASSERT(ValueIsAligned(partial_block->state.offset_in_block));
+	return true;
+}
+
+void PartialBlockManager::RegisterPartialBlock(PartialBlockAllocation &&allocation) {
+	auto &state(allocation.partial_block->state);
+	if (state.block_use_count < max_use_count) {
+		auto new_size = AlignValue(allocation.allocation_size + state.offset_in_block);
+		state.offset_in_block = new_size;
+		auto new_space_left = state.block_size - new_size;
+		// check if the block is STILL partially filled after adding the segment_size
+		if (new_space_left >= Storage::BLOCK_SIZE - free_space_threshold) {
```

Comment:
The original implementation was (new_space_left >=  free_space_threshold)

free_space_threshold is initialized to Storage::BLOCK_SIZE/5. So, the original implementation would reuse blocks that are at least 20% free. This code seems to only reuse blocks that are at least 80% free.

Is this the intent?

Reply:
Oh, I see. DEFAULT_FREE_SPACE_THRESHOLD is defined as 4*Storage::BLOCK_SIZE/5. I think that was a mistake.

I would propose changing this back to (new_space_left >= free_space_threshold) and redefining DEFAULT_FREE_SPACE_THRESHOLD as (Storage::BLOCK_SIZE/5).

Reply:
I redefined `free_space_threshold` back to the original definition of "the maximum size at which we consider a block to have free space", i.e. `80%` of a block size. This checks if the free space left in the block is larger than (100%-80% = 20%).

Perhaps the name `free_space_threshold` is confusing? Maybe `max_partial_block_size` is more clear?